system/src/Grav/Common/Page/Medium/ImageMedium.php

```diff
@@ -232,31 +218,36 @@ public function srcset($reset = true)
      * @return $this
      */
     public function derivatives($min_width, $max_width, $step = 200) {
-      $width = $min_width;
-
-      // Do not upscale images.
-      if ($max_width > $this->get('width')) {
-        $max_width = $this->get('width');
-      }
-
-      while ($width <= $max_width) {
-        $ratio = $width / $this->get('width');
-        $derivative = MediumFactory::scaledFromMedium($this, 1, $ratio);
-        if (is_array($derivative)) {
-          $this->addDerivative($derivative['file']);
+        $width = $min_width;
+
+        // Do not upscale images.
+        if ($max_width > $this->get('width')) {
+            $max_width = $this->get('width');
         }
-        $width += $step;
-      }
-      return $this;
-    }
 
-    /**
-     * Add a derivative
-     *
-     * @param  ImageMedium $image
-     */
-    public function addDerivative(ImageMedium $image) {
-      $this->derivatives[$image->url()] = $image->get('width');
+        // Clear any alternatives that have already been implied by the image's
+        // filename (eg. ones that had "@2x" or similar appended to their
+        // filenames)
+        $this->reset();
+
+        while ($width <= $max_width) {
+            $ratio = $width / $this->get('width');
+            $width += $step;
+
+            if ($ratio == 1) {
+                continue;
+            }
+
+            $derivative = MediumFactory::fromFile($this->get('filepath'));
+            $derivative_width = $derivative->get('width') * $ratio;
```

Comment:
I've been using this fork for my own project, and I've run across an error on this row a few times. `MediumFactory::fromFile` seems to have returned null, presumably because `$this->get('filepath')` in turn returned null. Both times this happened was after I had removed content, and the problem went away by restarting the `php-fpm` service on the server, so I suspect Grav has somehow cached the content associated with a page and then runs into an error.

Does this seem like a separate issue, or do we need to check to ensure the filepath isn't null or something here?